activesupport/lib/active_support/cache/mem_cache_store.rb

```diff
@@ -53,16 +53,18 @@ def self.supports_cache_versioning?
       ESCAPE_KEY_CHARS = /[\x00-\x20%\x7F-\xFF]/n
 
       # Creates a new Dalli::Client instance with specified addresses and options.
-      # By default address is equal localhost:11211.
+      # If no addresses are provided, we give nil to Dalli::Client, so it uses its fallbacks:
+      # - ENV["MEMCACHE_SERVERS"] (if defined)
+      # - "127.0.0.1:11211"        (otherwise)
       #
       #   ActiveSupport::Cache::MemCacheStore.build_mem_cache
-      #     # => #<Dalli::Client:0x007f98a47d2028 @servers=["localhost:11211"], @options={}, @ring=nil>
+      #     # => #<Dalli::Client:0x007f98a47d2028 @servers=["127.0.0.1:11211"], @options={}, @ring=nil>
       #   ActiveSupport::Cache::MemCacheStore.build_mem_cache('localhost:10290')
       #     # => #<Dalli::Client:0x007f98a47b3a60 @servers=["localhost:10290"], @options={}, @ring=nil>
       def self.build_mem_cache(*addresses) # :nodoc:
         addresses = addresses.flatten
         options = addresses.extract_options!
-        addresses = ["localhost:11211"] if addresses.empty?
+        addresses = nil if addresses.empty?
```

Comment:
Users of `:dalli_store` may have been passing an explicit `nil` parameter for the servers:

```ruby
config.cache_store = :dalli_cache, nil, { expires_in: 2.hour, compress: true }
```

If they simply changed `:dalli_cache` and `:mem_cache_store`, this code passes `addresses = [nil]` to Dalli (instead of `nil`, which cause exceptions when people try to access the cache.

```
NoMethodError: undefined method `match' for nil:NilClass
```

I've seen at least 2 cases within repos where this was the case, which confused hapless developers who were just trying to migrate in order to clear the deprecation warning (causing them to revert instead of debugging the issue).

Perhaps we could do something like add a `.compact`:

```ruby
addresses = nil if addresses.compact.empty?
```

This would unify the interface between the two stores.

Or perhaps raise an exception that explained what they were doing wrong / the difference in interface between `:dalli_store` and `:mem_cache_store`.

Reply:
I think it makes sense to explain, but I think what needs to explain how the migration should be done is the dalli_store project.

Reply:
Of some relevance, I opened https://github.com/petergoldstein/dalli/pull/764 a while back to document the change in implicit ENV variable handling.

Reply:
I also think the `compact` version is fine. @movermeyer want to open a PR?

Reply:
@rafaelfranca It would be my pleasure!

Reply:
https://github.com/rails/rails/pull/41381